# Textual Analysis: Text Pre-Processing and Cosine Similarity

## 1. Get business descriptions from Yahoo Finance website

## Intel Business Description


In [ ]:
intc_text = 'Intel Corporation designs, develops, manufactures, markets, and sells computing and related products worldwide. It operates through Client Computing Group, Data Center and AI, Network and Edge, Mobileye, Accelerated Computing Systems and Graphics, Intel Foundry Services, and Other segments. The company offers platform products, such as central processing units and chipsets, and system-on-chip and multichip packages; and accelerators, boards and systems, connectivity products, and memory and storage products. It also provides high-performance compute solutions for targeted verticals and embedded applications for retail, industrial, and healthcare markets; and solutions for assisted and autonomous driving comprising compute platforms, computer vision and machine learning-based sensing, mapping and localization, driving policy, and active sensors. In addition, the company offers workload-optimized platforms and related products for cloud service providers, enterprise and government, and communications service providers. It serves original equipment manufacturers, original design manufacturers, cloud service providers, and other equipment manufacturers. The company was incorporated in 1968 and is headquartered in Santa Clara, California.'
intc_text

'Intel Corporation designs, develops, manufactures, markets, and sells computing and related products worldwide. It operates through Client Computing Group, Data Center and AI, Network and Edge, Mobileye, Accelerated Computing Systems and Graphics, Intel Foundry Services, and Other segments. The company offers platform products, such as central processing units and chipsets, and system-on-chip and multichip packages; and accelerators, boards and systems, connectivity products, and memory and storage products. It also provides high-performance compute solutions for targeted verticals and embedded applications for retail, industrial, and healthcare markets; and solutions for assisted and autonomous driving comprising compute platforms, computer vision and machine learning-based sensing, mapping and localization, driving policy, and active sensors. In addition, the company offers workload-optimized platforms and related products for cloud service providers, enterprise and government, and 

## NVIDIA Business Description



In [ ]:
nvidia_text = "NVIDIA Corporation provides graphics, and compute and networking solutions in the United States, Taiwan, China, and internationally. The company's Graphics segment offers GeForce GPUs for gaming and PCs, the GeForce NOW game streaming service and related infrastructure, and solutions for gaming platforms; Quadro/NVIDIA RTX GPUs for enterprise workstation graphics; vGPU software for cloud-based visual and virtual computing; automotive platforms for infotainment systems; and Omniverse software for building 3D designs and virtual worlds. Its Compute & Networking segment provides Data Center platforms and systems for AI, HPC, and accelerated computing; Mellanox networking and interconnect solutions; automotive AI Cockpit, autonomous driving development agreements, and autonomous vehicle solutions; cryptocurrency mining processors; Jetson for robotics and other embedded platforms; and NVIDIA AI Enterprise and other software. The company's products are used in gaming, professional visualization, datacenter, and automotive markets. NVIDIA Corporation sells its products to original equipment manufacturers, original device manufacturers, system builders, add-in board manufacturers, retailers/distributors, independent software vendors, Internet and cloud service providers, automotive manufacturers and tier-1 automotive suppliers, mapping companies, start-ups, and other ecosystem participants. It has a strategic collaboration with Kroger Co. NVIDIA Corporation was incorporated in 1993 and is headquartered in Santa Clara, California."
nvidia_text

"NVIDIA Corporation provides graphics, and compute and networking solutions in the United States, Taiwan, China, and internationally. The company's Graphics segment offers GeForce GPUs for gaming and PCs, the GeForce NOW game streaming service and related infrastructure, and solutions for gaming platforms; Quadro/NVIDIA RTX GPUs for enterprise workstation graphics; vGPU software for cloud-based visual and virtual computing; automotive platforms for infotainment systems; and Omniverse software for building 3D designs and virtual worlds. Its Compute & Networking segment provides Data Center platforms and systems for AI, HPC, and accelerated computing; Mellanox networking and interconnect solutions; automotive AI Cockpit, autonomous driving development agreements, and autonomous vehicle solutions; cryptocurrency mining processors; Jetson for robotics and other embedded platforms; and NVIDIA AI Enterprise and other software. The company's products are used in gaming, professional visualiza

## 2. Provide some data pre-processing: decapitalize the text, remove punctuation and other unwanted characters.

---
re.compile() → compiles a regular expression pattern into an object.

'<.*?>' → pattern that matches HTML tags:

<br>

< and > → start and end of a tag

<br>

.*? → matches any characters in between, as few as possible

<br>

This creates a rule cleanr to identify HTML tags.

<br>

re.sub(pattern, replacement, string) → replaces text based on the pattern

<br>

pattern → what to look for (cleanr)

<br>

replacement → replace with this (here '' means remove)

<br>

string → the text to search in (raw_html)

<br>

This line removes all HTML tags from raw_html, resulting in cleantext.

---

In [ ]:
#STEP 1: Remove html tags
#Define the function to remove html tags
import re

def cleanhtml(raw_html):
  cleanr = re.compile('<.*?>')
  cleantext = re.sub(cleanr, '', raw_html)
  return cleantext

In [ ]:
cleantext_intc= cleanhtml(intc_text)
cleantext_intc

'Intel Corporation designs, develops, manufactures, markets, and sells computing and related products worldwide. It operates through Client Computing Group, Data Center and AI, Network and Edge, Mobileye, Accelerated Computing Systems and Graphics, Intel Foundry Services, and Other segments. The company offers platform products, such as central processing units and chipsets, and system-on-chip and multichip packages; and accelerators, boards and systems, connectivity products, and memory and storage products. It also provides high-performance compute solutions for targeted verticals and embedded applications for retail, industrial, and healthcare markets; and solutions for assisted and autonomous driving comprising compute platforms, computer vision and machine learning-based sensing, mapping and localization, driving policy, and active sensors. In addition, the company offers workload-optimized platforms and related products for cloud service providers, enterprise and government, and 

In [ ]:
cleantext_nvidia= cleanhtml(nvidia_text)
cleantext_nvidia

"NVIDIA Corporation provides graphics, and compute and networking solutions in the United States, Taiwan, China, and internationally. The company's Graphics segment offers GeForce GPUs for gaming and PCs, the GeForce NOW game streaming service and related infrastructure, and solutions for gaming platforms; Quadro/NVIDIA RTX GPUs for enterprise workstation graphics; vGPU software for cloud-based visual and virtual computing; automotive platforms for infotainment systems; and Omniverse software for building 3D designs and virtual worlds. Its Compute & Networking segment provides Data Center platforms and systems for AI, HPC, and accelerated computing; Mellanox networking and interconnect solutions; automotive AI Cockpit, autonomous driving development agreements, and autonomous vehicle solutions; cryptocurrency mining processors; Jetson for robotics and other embedded platforms; and NVIDIA AI Enterprise and other software. The company's products are used in gaming, professional visualiza

In [ ]:
#Remove punctuation
# initializing punctuations string
punc = '''!()-[]{};:'"\,<>./?@#$%^&*_~'''

# Removing punctuations in string
# Using loop + punctuation string
for ele in cleantext_intc:
    if ele in punc:
        cleantext_intc = cleantext_intc.replace(ele, "")
cleantext_intc

'Intel Corporation designs develops manufactures markets and sells computing and related products worldwide It operates through Client Computing Group Data Center and AI Network and Edge Mobileye Accelerated Computing Systems and Graphics Intel Foundry Services and Other segments The company offers platform products such as central processing units and chipsets and systemonchip and multichip packages and accelerators boards and systems connectivity products and memory and storage products It also provides highperformance compute solutions for targeted verticals and embedded applications for retail industrial and healthcare markets and solutions for assisted and autonomous driving comprising compute platforms computer vision and machine learningbased sensing mapping and localization driving policy and active sensors In addition the company offers workloadoptimized platforms and related products for cloud service providers enterprise and government and communications service providers It

In [ ]:
# Removing punctuations in string
# Using loop + punctuation string
for ele in cleantext_nvidia:
    if ele in punc:
        cleantext_nvidia = cleantext_nvidia.replace(ele, "")
cleantext_nvidia

'NVIDIA Corporation provides graphics and compute and networking solutions in the United States Taiwan China and internationally The companys Graphics segment offers GeForce GPUs for gaming and PCs the GeForce NOW game streaming service and related infrastructure and solutions for gaming platforms QuadroNVIDIA RTX GPUs for enterprise workstation graphics vGPU software for cloudbased visual and virtual computing automotive platforms for infotainment systems and Omniverse software for building 3D designs and virtual worlds Its Compute  Networking segment provides Data Center platforms and systems for AI HPC and accelerated computing Mellanox networking and interconnect solutions automotive AI Cockpit autonomous driving development agreements and autonomous vehicle solutions cryptocurrency mining processors Jetson for robotics and other embedded platforms and NVIDIA AI Enterprise and other software The companys products are used in gaming professional visualization datacenter and automoti

In [ ]:
#Step 3: Decapitalization
cleantext_intc = cleantext_intc.lower()
cleantext_intc

'intel corporation designs develops manufactures markets and sells computing and related products worldwide it operates through client computing group data center and ai network and edge mobileye accelerated computing systems and graphics intel foundry services and other segments the company offers platform products such as central processing units and chipsets and systemonchip and multichip packages and accelerators boards and systems connectivity products and memory and storage products it also provides highperformance compute solutions for targeted verticals and embedded applications for retail industrial and healthcare markets and solutions for assisted and autonomous driving comprising compute platforms computer vision and machine learningbased sensing mapping and localization driving policy and active sensors in addition the company offers workloadoptimized platforms and related products for cloud service providers enterprise and government and communications service providers it

In [ ]:
cleantext_nvidia = cleantext_nvidia.lower()
cleantext_nvidia

'nvidia corporation provides graphics and compute and networking solutions in the united states taiwan china and internationally the companys graphics segment offers geforce gpus for gaming and pcs the geforce now game streaming service and related infrastructure and solutions for gaming platforms quadronvidia rtx gpus for enterprise workstation graphics vgpu software for cloudbased visual and virtual computing automotive platforms for infotainment systems and omniverse software for building 3d designs and virtual worlds its compute  networking segment provides data center platforms and systems for ai hpc and accelerated computing mellanox networking and interconnect solutions automotive ai cockpit autonomous driving development agreements and autonomous vehicle solutions cryptocurrency mining processors jetson for robotics and other embedded platforms and nvidia ai enterprise and other software the companys products are used in gaming professional visualization datacenter and automoti

## 3. Calculate cosine similarity between these two texts.

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
#combine 2 texts to list
text=[]
all = [cleantext_intc,cleantext_nvidia]
for i in all:
  text.append(i)
text

['intel corporation designs develops manufactures markets and sells computing and related products worldwide it operates through client computing group data center and ai network and edge mobileye accelerated computing systems and graphics intel foundry services and other segments the company offers platform products such as central processing units and chipsets and systemonchip and multichip packages and accelerators boards and systems connectivity products and memory and storage products it also provides highperformance compute solutions for targeted verticals and embedded applications for retail industrial and healthcare markets and solutions for assisted and autonomous driving comprising compute platforms computer vision and machine learningbased sensing mapping and localization driving policy and active sensors in addition the company offers workloadoptimized platforms and related products for cloud service providers enterprise and government and communications service providers i

In [ ]:
type(text)

list

In [ ]:
#Using sklearn vectorize the documents:
from sklearn.feature_extraction.text import CountVectorizer

# Create a matrix to represent the corpus
X = CountVectorizer().fit_transform(text).toarray()

print(X)

[[ 1  0  0  1  1  1  0  1  0  1  1 25  1  0  1  1  0  1  0  1  0  0  1  1
   1  0  1  1  1  2  0  0  0  0  1  0  3  0  1  2  1  3  1  1  0  1  0  1
   1  0  1  0  2  0  1  1  1  2  4  1  0  0  0  1  0  1  1  0  1  1  1  0
   3  1  0  1  0  0  2  0  0  0  1  3  0  0  0  1  1  1  3  1  1  2  0  1
   0  1  1  1  0  0  0  2  0  1  2  2  1  0  0  1  2  1  1  0  5  0  3  1
   0  2  1  0  0  0  1  0  1  1  1  1  1  3  1  0  2  0  0  1  0  0  1  0
   0  1  2  0  1  3  1  0  0  0  1  0  0  0  1  0  0  1  0  0  1  0  1  0
   0  1]
 [ 0  1  1  1  0  0  1  0  1  3  0 21  0  1  0  0  5  2  1  0  1  1  1  1
   0  1  0  1  0  1  1  1  1  1  0  1  0  2  0  2  0  2  0  3  1  1  1  0
   1  1  0  1  1  1  0  1  2  1  8  0  1  3  2  0  2  3  0  1  1  0  0  1
   4  1  1  0  1  1  0  1  1  1  1  1  2  1  1  0  0  0  4  0  1  1  1  0
   1  0  0  0  3  1  4  1  1  0  2  3  0  1  1  0  4  0  0  1  2  1  1  2
   1  1  0  1  1  1  1  2  0  1  0  0  0  2  0  4  4  1  1  0  1  1  0  1
   1  0  2  1  0  4  0  1  1 

In [ ]:
#Or we can use the following code, which removes stop words and shows the dataframe with the word frequencies
# Scikit Learn
from sklearn.feature_extraction.text import CountVectorizer
import pandas as pd

# Create the Document Term Matrix
count_vectorizer = CountVectorizer(stop_words='english')
count_vectorizer = CountVectorizer()
sparse_matrix = count_vectorizer.fit_transform(text)

# OPTIONAL: Convert Sparse Matrix to Pandas Dataframe if you want to see the word frequencies.
doc_term_matrix = sparse_matrix.todense()
df = pd.DataFrame(doc_term_matrix,
                  columns=count_vectorizer.get_feature_names_out(),
                  index=['cleantext_intc', 'cleantext_nvidia'])
df

1968  1993  3d  accelerated  accelerators  active  addin  \
cleantext_intc       1     0   0            1             1       1      0   
cleantext_nvidia     0     1   1            1             0       0      1   

                  addition  agreements  ai  ...  virtual  vision  visual  \
cleantext_intc           1           0   1  ...        0       1       0   
cleantext_nvidia         0           1   3  ...        2       0       1   

                  visualization  was  with  workloadoptimized  workstation  \
cleantext_intc                0    1     0                  1            0   
cleantext_nvidia              1    1     1                  0            1   

                  worlds  worldwide  
cleantext_intc         0          1  
cleantext_nvidia       1          0  

[2 rows x 170 columns]

## Creating a Readable Document-Term DataFrame

- **doc_term_matrix**: The actual numbers showing how many times each word appears in each document.  
- **columns=cv.get_feature_names_out()**: The column names are the words themselves, so you know which column corresponds to which word.  
- **index=['cleantext_intc', 'cleantext_nvidia']**: The row names are the labels for the documents, so you know which row corresponds to Intel and which to Nvidia.  

<br>

This converts a plain numeric matrix into a readable table, showing word counts for each document.

---

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity

cosine_similarity(df, df)

array([[1.        , 0.81746918],
       [0.81746918, 1.        ]])

##4. Brief summary of the level of similarity. Are there significant relation between under/overvaluation (if present) and cosine similarity?

Cosine similarity is a metric used to determine how similar documents are
irrespective of their size. Mathematically, it measures the cosine of the angle between two vectors projected in a multi-dimensional space. The two vectors are arrays containing the word and word counts of two
documents.

<br>

Since the cosine similarity value is 0.817, we can say that the two texts are similar. The description of Intel and NVIDIA are very similar, this is true because both of these companies belong to the same industry.

<br>

Yes, we can see a significant relation between overvaluation and cosine similarity. Both stocks were Over-valued and both stocks have higher similarity in the type of business they belong to.

---

## Part 2 Valuation: Calculation of WACC

 1. Calculate WACC based on financial information for 2022. To calculate cost of Debt I can look into 10-K Notes about Debt, or I can take the cost of debt from analysts reports/ fin websites. For Cost of Equity calculation I use the Equity Risk Premium and Risk free rate from Damodaran website and from Fred website

 WACC = Cost of equity * (Equity/(Equity + Debt)) + Cost of debt * (Debt/(Equity + Debt))(1 - tax)

## INTEL WACC Calculation

**Cost of Debt** from fin website:

https://valueinvesting.io/INTC/valuation/wacc

Cost of Debt for Intel = 5.5%

<br>

**Cost of Equity related data**

From Damodaran website:

https://pages.stern.nyu.edu/~adamodar/New_Home_Page/datafile/ctryprem.html

United States:

Equity Risk Premium = 5.94%

Country Risk Premium = 0.00%

Corporate Tax Rate = 25.00%

Moody's rating = Aaa

<br>

**Market Yield on U.S. Treasury Securities at 20-Year Constant Maturity, Quoted on an Investment Basis from website**:

https://fred.stlouisfed.org/series/DGS20

Risk free rate = 4.14% as on 30. December. 2022

<br>

**Cost of Equity** = Risk free rate + Beta * Market Risk premium + Country Risk Premium

<br>

**Year-End Beta** from website:

https://wrds-www.wharton.upenn.edu/

WRDS CRSP -> Portfolio Assignments -> Year-end Beta Value

Beta as on 30. December. 2022 = 0.79224

Cost of Equity = 4.14% + 0.79224 * 5.94% + 0.00%

Cost of Equity = 0.08845 = 8.845%

<br>

**Debt and Equity** from website:

https://finance.yahoo.com/quote/INTC/balance-sheet?p=INTC

Total Debt = 42,051,000 as on 30. December. 2022

Total Equity = 101,423,000 as on 30. December. 2022

<br>

Now, **WACC** is:

WACC = Cost of equity * (Equity/(Equity + Debt)) + Cost of debt * (Debt/(Equity + Debt))(1 - tax)

Weight of Debt = 29.31%

Weight of equity = 70.69%

**WACC** = 8.845% * (101,423,000/(101,423,000 + 42,051,000)) + 5.5% * (42,051,000/(101,423,000 + 42,051,000))(1 - 25.00%)

<br>

**WACC of Intel = 7.46%**

---
### How to Determine Shareholders' Expected Return (Cost of Equity) and Stock Fluctuations

**1️⃣ Determining Cost of Equity**

Shareholders' expected return is not random; it can be estimated using the **CAPM model**:

<br>

**Explanation:**  
- Shareholders want a reasonable return **for taking on risk**.  
- CAPM quantifies this expected return.  

<br>
---
<br>

**2️⃣ Stock Price Fluctuations vs. Expected Return**

- Stocks fluctuate daily, but shareholders care about **long-term average return**, not daily movements.  
- Short-term price drops do not mean shareholders will lose over the long run.  
- As long as the long-term average return meets or exceeds the **Cost of Equity**, shareholders are satisfied.  

<br>

**Summary:**  
- Shareholder expected return = long-term average return  
- Daily stock fluctuations ≠ expected return  
- CAPM provides a quantified target for reasonable shareholder expectations

---


## NVIDIA WACC Calculation

**Cost of Debt** from fin website:

https://valueinvesting.io/NVDA/valuation/wacc

Cost of Debt for NVIDIA = 4.45%

<br>

**Cost of Equity related data**

From Damodaran website:

https://pages.stern.nyu.edu/~adamodar/New_Home_Page/datafile/ctryprem.html

United States:

Equity Risk Premium = 5.94%

Country Risk Premium = 0.00%

Corporate Tax Rate = 25.00%

Moody's rating = Aaa

<br>

**Market Yield on U.S. Treasury Securities at 20-Year Constant Maturity, Quoted on an Investment Basis from website**:

https://fred.stlouisfed.org/series/DGS20

Risk free rate = 4.14% as on 30. December. 2022

<br>

**Cost of Equity** = Risk free rate + Beta * Market Risk premium + Country Risk Premium

**Year-End Beta** from website:

https://wrds-www.wharton.upenn.edu/

WRDS CRSP -> Portfolio Assignments -> Year-end Beta Value

Beta as on 30. December. 2022 = 1.82711

Cost of Equity = 4.14% + 1.82711 * 5.94% + 0.00%

Cost of Equity = 0.1499 = 14.99%

<br>

**Debt and Equity** from website:

https://finance.yahoo.com/quote/NVDA/balance-sheet?p=NVDA

Total Debt = 11,855,000 as on 30. December. 2022

Total Equity = 22,101,000 as on 30. December.

<br>

Now, **WACC** is:

WACC = Cost of equity * (Equity/(Equity + Debt)) + Cost of debt * (Debt/(Equity + Debt))(1 - tax)

Weight of Debt = 34.91%

Weight of equity = 65.09%

**WACC** = 14.99% * 65.09% + 4.45% * 34.91% * (1 - 25.00%)

<br>

**WACC of NVIDIA = 10.92%**

---
##WACC Overview

WACC tells you: for every $1 the company uses from the stockers, how much it costs on average.

<br>

##  Significance of WACC

### Investment Decision Standard
- If the expected return of a project > WACC → the project adds value  
- If < WACC → the project destroys value  

<br>

### Valuation Reference
- In DCF (Discounted Cash Flow) models, WACC is used to discount future cash flows  
- Lower WACC → higher present value of future cash flows  

<br>

### Measure of Capital Structure Efficiency
- WACC reflects whether the company's financing mix is reasonable  
- Example: if the company has more debt with low interest, WACC may be lower than a company with mostly equity

---

## 2. Comparison of WACCs

WACC is weighted average cost of capital and cost of debt.

WACC of Intel = 7.46%

WACC of NVIDIA = 10.92%

WACC of NVIDIA is higher than that of Intel.

---

## 3. Assume that the projected Cash Flows for two companies will be totally the same, which company would I invest in and why?

From the perspective of WACC:

We need to discount Free Cash Flows using WACC

**Market Value of the Company = Future Cash Flows / WACC**

<br>

Companies should decrease WACC because their main aim is to maximize
shareholder Value. So, if the projected Cash Flows for two companies will be totally the same, I would invest in **Intel** company since it has a lower WACC, therefore it will be a better investment as an investor.

---

### 4. In reality the Cash Flows for these two companies will differ. Which company in my opinion will have higher growth rate of Cash flows? Why do I think so? Explain which growth rate projection approach would you choose?

## Factors to Estimate Future Cash Flow Growth

1. **Historical Growth Trend**  
   Look at the company's cash flow growth over the past few years.  
   *Example:* If Intel's cash flow grew 5% per year over the last three years, while NVIDIA's grew 15%, NVIDIA may have higher short-term growth potential.

2. **Industry & Market Outlook**  
   Consider the growth rate of the industry the company operates in.  
   Fast-growing industries (e.g., AI, cloud computing, new semiconductor tech) usually lead to faster cash flow growth.  
   Mature, slow-growth industries usually have stable but lower cash flow growth.

3. **Business Strategy & Competitive Advantage**  
   R&D investment, product innovation, and market share expansion can increase future cash flow growth.  
   Companies with a strong moat can maintain higher profits and cash flow.

4. **Financial Health**  
   High debt can limit expansion and restrict cash flow growth.  
   Companies with high profit margins and strong free cash flow can reinvest and expand business to drive future cash flow growth.

5. **Growth Rate Projection Methods**  
   - **Historical CAGR Method:** Calculate compound annual growth rate based on past cash flows; suitable for stable companies.  
   - **Analyst Estimates:** Use forecasts from financial websites or research reports for future years.  
   - **Multi-stage Growth Model:** Assume faster growth in early years (e.g., first 5 years) and slower growth as the company matures.

<br>

---

<br>

##  Types of Cash Flow

1. **Operating Cash Flow (OCF)**  
   Cash generated from daily business operations, e.g., selling products or providing services minus daily expenses (wages, materials).  
   *Example:* Selling computers earns 1,000,000, pay 700,000 in wages and materials → OCF = 300,000.

2. **Investing Cash Flow (ICF)**  
   Cash spent or received from buying/selling long-term assets, e.g., machinery, property, or investments.  
   *Example:* Buying a new machine costs 200,000 → ICF = -200,000.

3. **Financing Cash Flow (FCF)**  
   Cash from external financing or repayments, e.g., borrowing, repaying debt, issuing stock, or paying dividends.  
   *Example:* Borrow 500,000 from a bank → FCF = +500,000; pay 100,000 in dividends → FCF = -100,000.

<br>

---

<br>

Opinion:

In my opinion, NVIDIA will have higher growth rate of cash flows.

<br>

I think so because of the last few years performance. I would choose the "Analysis of past dynamics" growth rate projection approach. We can look at historical growth of earnings per share (EPS), revenue, net income, etc.

<br>

In the year 2022, Intel had negative Free Cash Flow while NVIDIA had positive Free Cash flow. Though both of them reported declines in 2020, the extent of decline in INTEL Cash flow was much worse.


<br>

References:

https://finance.yahoo.com/quote/INTC/cash-flow?p=INTC

https://finance.yahoo.com/quote/NVDA/cash-flow?p=NVDA
